runtime/calamari/src/lib.rs

```diff
@@ -922,8 +923,24 @@ pub type Executive = frame_executive::Executive<
 	frame_system::ChainContext<Runtime>,
 	Runtime,
 	AllPalletsReversedWithSystemFirst,
+	CollatorSelectionMigrationV2,
 >;
 
+pub struct CollatorSelectionMigrationV2;
+impl OnRuntimeUpgrade for CollatorSelectionMigrationV2 {
+	fn on_runtime_upgrade() -> Weight {
+		CollatorSelection::migrate_to_v2()
+	}
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<(), &'static str> {
+		CollatorSelection::pre_migrate_to_v2()
+	}
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade() -> Result<(), &'static str> {
+		CollatorSelection::post_migrate_to_v2()
+	}
+}
+
```

Comment:
You can also implement `OnRuntimeUpgrade` for some struct, feature-gated by `try-runtim`.
In the method implementations you can write to temporary storage and check invariants before and after any runtime upgrade.
Something simple to check is that collators still exist and are still the same accounts.
https://www.crowdcast.io/e/substrate-seminar/41 at 1hour 15min gives an example

Reply:
I tried having at least the pre/post_upgrade hooks gated by `try-runtime`, but then the runtime wouldn't compile.
I don't think the `on_runtime_upgrade hook can't be featuregated

Reply:
I mean exactly what you did for `CollatorSelectionMigrationV2 ` but you can read the collators `pre_upgrade`, write them to temp storage and then read the temp storage in `post_upgrade` and compare to what is in the storage then, to make sure they are the same.

Reply:
updated

Reply:
I think the "upgrade checker" migrations are worthy of their own issue. I added one: #463
I also added a check that the expected migrations were executed